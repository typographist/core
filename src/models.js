// @flow

// userConfig models
export type UserConfigBreakpoint = {
  value: string,
  base?: string | string[],
  lineHeight?: number,
  ratio?: number,
};

export type UserConfig = {
  base: string | string[],
  lineHeight: number,
  ratio: string | number,
  tablet: UserConfigBreakpoint,
  desktop: UserConfigBreakpoint,
  lgDesktop: UserConfigBreakpoint,
  xlDesktop: UserConfigBreakpoint,
};

export type NotFilledBreakpoint = {
  value: string,
  base?: number[],
  lineHeight?: number,
  ratio?: number | string,
  name: string,
};

export type FilledBreakpoint = {|
  value: string,
  base: number[],
  lineHeight: number,
  ratio: number | string,
  name: string,
|};

export type FilledBreakpointStableRatio = {|
  value: string,
  base: number[],
  lineHeight: number,
  ratio: number,
  name: string,
|};

export type Breakpoint = {
  value: string,
  base: number[],
  lineHeight: number,
  ratio: number,
  root: number,
  name: string,
};
