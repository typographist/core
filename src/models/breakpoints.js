// @flow

export type NotFilledBreakpoint = {
  breakpoint: string,
  name: string,
  base?: number[],
  lineHeight?: number,
  ratio?: number | string,
};

export type NamedBreakpoint = {
  breakpoint: string,
  name: string,
  base?: string | string[],
  lineHeight?: number,
  ratio?: number | string,
};

export type FilledBreakpoint = {|
  value: string,
  base: number[],
  lineHeight: number,
  ratio: number | string,
  name: string,
|};

export type FilledBreakpointStableRatio = {|
  value: string,
  base: number[],
  lineHeight: number,
  ratio: number,
  name: string,
|};

export type Breakpoint = {
  value: string,
  base: number[],
  lineHeight: number,
  ratio: number,
  root: number,
  name: string,
};

export const breakpoints: Breakpoint[] = [
  {
    base: [16, 32],
    lineHeight: 1.5,
    name: 'default',
    ratio: 1.1880883987824906,
    root: 12,
    value: '0px',
  },
  {
    base: [17],
    lineHeight: 1.5,
    name: 'tablet',
    ratio: 1.1761442744249144,
    root: 13,
    value: '640px',
  },
  {
    base: [18],
    lineHeight: 1.7,
    name: 'desktop',
    ratio: 1.333,
    root: 16,
    value: '1024px',
  },
  {
    base: [20],
    lineHeight: 1.7,
    name: 'lgDesktop',
    ratio: 1.333,
    root: 17,
    value: '1200px',
  },
  {
    base: [22],
    lineHeight: 1.7,
    name: 'xlDesktop',
    ratio: 1.333,
    root: 19,
    value: '1600px',
  },
];

export type BreakpointWithoutName = {
  value: string,
  base: number[],
  lineHeight: number,
  ratio: number,
  root: number,
};

type BreakpointsMap = {
  [string]: BreakpointWithoutName,
};

export const breakpointsMap: BreakpointsMap = {
  default: {
    base: [16, 32],
    lineHeight: 1.5,
    ratio: 1.1880883987824906,
    root: 12,
    value: '0px',
  },
  tablet: {
    base: [17],
    lineHeight: 1.5,
    ratio: 1.1761442744249144,
    root: 13,
    value: '640px',
  },
  desktop: {
    base: [18],
    lineHeight: 1.7,
    ratio: 1.333,
    root: 16,
    value: '1024px',
  },
  lgDesktop: {
    base: [20],
    lineHeight: 1.7,
    ratio: 1.333,
    root: 17,
    value: '1200px',
  },
  xlDesktop: {
    base: [22],
    lineHeight: 1.7,
    ratio: 1.333,
    root: 19,
    value: '1600px',
  },
};
