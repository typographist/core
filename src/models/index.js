//  @flow

export * from './breakpoints';
export * from './user-config';
