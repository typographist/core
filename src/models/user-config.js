// @flow

export type UserConfig = {
  base: string | string[],
  lineHeight: number,
  ratio: string | number,
  [string]: ?{
    breakpoint: string,
    base?: string | string[],
    lineHeight?: number,
    ratio?: number,
  },
};
