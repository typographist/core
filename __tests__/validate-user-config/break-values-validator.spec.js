import { userConfig } from '../../src/models/user-config';
import {
  getBreakValues,
  hasPxOrEm,
  isValidBreakValue,
  eitherIsValidField,
  isValidField,
  validateFields,
} from '../../src/validate-user-config/break-values-validator';

describe('getBreakValues function', () => {
  it('returs all values of breakpoint from user config', () => {
    expect(getBreakValues(userConfig)).toEqual([
      '40em',
      '64em',
      '75em',
      '100em',
    ]);
  });
});

describe('hasPxOrEm', () => {
  it("returns `false` if value doesn't contain px or em", () => {
    expect(hasPxOrEm(12)).toEqual(false);
  });

  it('returns `true` if value contains pixels', () => {
    expect(hasPxOrEm('12px')).toEqual(true);
  });

  it('returns `true` if value contained ems', () => {
    expect(hasPxOrEm('12em')).toEqual(true);
  });
});

describe('isValidBreakValue', () => {
  it('returns `false` if value is not string', () => {
    expect(isValidBreakValue(12)).toEqual(false);
  });

  it('returns `true` if value contains pixels', () => {
    expect(isValidBreakValue('12px')).toEqual(true);
  });

  it('returns `true` if value contains ems', () => {
    expect(isValidBreakValue('12em')).toEqual(true);
  });
});

describe('eitherIsValidField', () => {
  it('returns `true` breakpoint value contains pixels', () => {
    expect(eitherIsValidField('1200px').isRight).toEqual(true);
  });

  it('returns `true` if the breakpoint value has ems', () => {
    expect(eitherIsValidField('40em').isRight).toEqual(true);
  });

  it("returns `false` if the breakpoint does't contain px or ems ", () => {
    expect(eitherIsValidField('40rem').isLeft).toEqual(true);
  });
});

describe('isValidField', () => {
  it('returns `true` if value contain px', () => {
    expect(isValidField('12px')).toEqual(true);
  });

  it('returns `true` if value contain em', () => {
    expect(isValidField('12em')).toEqual(true);
  });
});

describe('validateFields function', () => {
  it('all breakpoints are valid', () => {
    expect(validateFields(userConfig)).toEqual(true);
  });
});
