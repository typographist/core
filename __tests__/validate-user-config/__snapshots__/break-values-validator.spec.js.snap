// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user config validators breakpoints validator breakValHasPxOrEm function returns \`true\` if the breakpoint value has ems 1`] = `"Typographist >>> Check your user config. 'breakpoint:  must be in pixels. For example breakpoint: '1024px'."`;
